File: 38.cpp
//
// Created by zb198 on 2019/9/22.
//
#include <vector>
#include <string>
#include <algorithm>
using namespace std;
class Solution {
private:
    vector<string> result;
public:
    void core(string str,int begin){
        if(begin==str.size()){
            auto Iter = find(result.begin(),result.end(),str);
            if(Iter == result.end()) {
                result.push_back(str);
            }
        } else{
            for(int Index=begin;Index<str.size();Index++){
                char temp = str[begin];
                str[begin] = str[Index];
                str[Index] = temp;
                core(str,begin+1);
                temp = str[begin];
                str[begin] = str[Index];
                str[Index] = temp;
            }
        }
    }
    vector<string> Permutation(string str) {
        if(str==""){
            return result;
        }
        core(str,0);
        sort(result.begin(),result.end());
        return result;
    }
};

File: 38.py
# -*- coding:utf-8 -*-
class Solution:
    def __init__(self):
        self.result = []
    def _core(self,ss:list,begin:int):
        if(begin==len(ss)):
            string = "".join(ss)
            if(string not in self.result):
                self.result.append(string)
        else:
            for index in range(begin,len(ss)):
                if(ss[index] != ss[begin]):
                    char = ss[index]
                    ss[index] = ss[begin]
                    ss[begin] = char
                self._core(ss,begin+1)
                if(ss[index] != ss[begin]):
                    char = ss[begin]
                    ss[begin] = ss[index]
                    ss[index] = char

    def Permutation(self, ss:str)->list:
        self._core([i for i in ss],0)
        return self.result
# write code here
if __name__ == '__main__':
    handly = Solution()
    result = handly.Permutation("aa")
    print(result)